File: internal/pythonfacts/module_imports.go
package pythonfacts

import (
	"encoding/json"
	"fmt"
	"slices"
)

type ModuleImport struct {
	Path    string `json:"path"`
	Kind    string `json:"kind"`
	Literal bool   `json:"literal"`
	ComputedImport
}

type moduleImportCall struct {
	Site      SourceLocation             `json:"site"`
	EndSite   SourceLocation             `json:"endSite"`
	Shape     string                     `json:"shape"`
	Arguments []objectImportCallArgument `json:"arguments"`
	Keywords  []json.RawMessage          `json:"keywords"`
}

func validateModuleImports(imports []ModuleImport, modules []TypeModule) error {
	inputs := map[objectImportCallSite]moduleImportCall{}
	for _, module := range modules {
		var facts struct {
			Calls []moduleImportCall `json:"calls"`
		}
		if err := json.Unmarshal(module.Facts, &facts); err != nil {
			return err
		}
		for _, call := range facts.Calls {
			inputs[objectImportCallSite{Path: module.Path, Site: call.Site, EndSite: call.EndSite}] = call
		}
	}
	seen := map[objectImportCallSite]bool{}
	for _, imported := range imports {
		key := objectImportCallSite{Path: imported.Path, Site: SourceLocation{Line: imported.Line, Column: imported.Column}, EndSite: SourceLocation{Line: imported.EndLine, Column: imported.EndColumn}}
		call, found := inputs[key]
		if !found || seen[key] || !validModuleImport(imported, call) {
			return fmt.Errorf("module import evidence does not match one current bounded source call")
		}
		seen[key] = true
	}
	return nil
}

func validModuleImport(imported ModuleImport, call moduleImportCall) bool {
	argument := objectImportCallArgument{}
	if len(call.Arguments) > 0 {
		argument = call.Arguments[0]
	}
	if !validModuleImportIdentity(imported, call, argument.Text) || !validModuleImportCollections(imported) {
		return false
	}
	if !validModuleImportScope(imported) {
		return false
	}
	if imported.EvidenceError != "" {
		return !imported.Literal && len(imported.Targets) == 0 && len(imported.Configuration) == 0 && imported.EntryPointGroup == ""
	}
	return validModuleImportTargets(imported, call, argument)
}

func validModuleImportIdentity(imported ModuleImport, call moduleImportCall, argument string) bool {
	return imported.Argument == argument && len(imported.Argument) <= 4096 && imported.Shape != "" && imported.Shape == call.Shape && len(imported.Shape) <= 16384 && imported.Callable != ""
}

func validModuleImportCollections(imported ModuleImport) bool {
	return imported.Targets != nil && imported.Configuration != nil && len(imported.Targets) <= 4096 && len(imported.Configuration) <= 4096 && len(imported.EvidenceError) <= 4096
}

func validModuleImportScope(imported ModuleImport) bool {
	return (imported.Callee == "importlib.import_module" || imported.Callee == "builtins.__import__") && (imported.Kind == "type-only" || imported.Kind == "proven-dynamic")
}

func validModuleImportTargets(imported ModuleImport, call moduleImportCall, argument objectImportCallArgument) bool {
	if len(call.Arguments) != 1 || len(call.Keywords) != 0 {
		return false
	}
	if imported.Literal {
		return argument.Kind == "string" && slices.Equal(imported.Targets, []string{argument.Value}) && len(imported.Configuration) == 0 && imported.EntryPointGroup == ""
	}
	return argument.Kind != "string" && len(imported.Targets)+len(imported.Configuration)+len(imported.EntryPointGroup) > 0
}

File: internal/pythonfacts/module_imports.py
from loader_bindings import LoaderBindings
from module_evidence import ModuleEvidence, module_name
from type_resolver import _Resolver

LOADERS = {"importlib.import_module", "builtins.__import__"}


def callable_names(module):
    names = {}
    for binding in module["facts"]["bindings"]:
        if binding["kind"] in {"function", "class"}:
            site = binding["site"]
            scope = (
                f"{binding['scope']}/{binding['kind']}:{site['line']}:{site['column']}"
            )
            names[scope] = binding["name"]
    result = {"module": "<module>"}
    for scope in module["scopes"]:
        if scope == "module":
            continue
        parent = result[scope.rpartition("/")[0]]
        name = names.get(
            scope, "<lambda>" if module["scopes"][scope]["kind"] == "function" else ""
        )
        result[scope] = (
            (parent + "." if parent != "<module>" and name else "") + name
            if name
            else parent
        )
    return result


def module_imports(modules, facts=None):
    facts = facts or _Resolver(modules)
    bindings = LoaderBindings(facts, ("__import__", "next"))
    result = []
    for path, module in sorted(facts.files.items()):
        evidence = ModuleEvidence(module, bindings)
        names = callable_names(module)
        calls = sorted(
            module["facts"]["calls"],
            key=lambda call: (call["site"]["line"], call["site"]["column"]),
        )
        for call in calls:
            if call["callee"]["kind"] != "name":
                continue
            candidates = bindings.candidates(
                module,
                call["scope"],
                call["callee"]["value"],
                call["site"],
                call["flow"],
            )
            loaders = candidates & LOADERS
            if not loaders:
                continue
            result.append(
                import_fact(
                    path,
                    call,
                    names[call["scope"]],
                    min(loaders),
                    candidates,
                    evidence,
                )
            )
    return result


def import_fact(path, call, callable_name, callee, candidates, resolver):
    arguments = call["arguments"]
    argument = arguments[0]["text"] if arguments else ""
    if not call["shape"] or len(argument.encode("utf-8")) > 4096:
        raise ValueError("Python module loader exceeds its bounded call evidence")
    error = ""
    evidence = None
    if candidates != {callee}:
        error = "module loader binding is ambiguous, conditional, or wildcard"
    elif len(arguments) != 1 or call["keywords"]:
        error = "module loader requires one positional argument"
    else:
        evidence = resolver.evidence(arguments[0])
        if evidence is None:
            error = "argument is outside the supported bounded computed-import shapes"
    targets, configuration, group = evidence or (set(), set(), "")
    type_only = path.endswith(".pyi") or any(
        resolver.bindings.reference(
            resolver.module, guard["scope"], guard["name"], guard["site"]
        )
        == "typing.TYPE_CHECKING"
        for guard in call["typeGuards"]
    )
    literal = (
        not error
        and arguments[0]["kind"] == "string"
        and module_name(arguments[0]["value"])
    )
    return {
        "path": path,
        "kind": "type-only" if type_only else "proven-dynamic",
        "literal": bool(literal),
        "callee": callee,
        "callable": callable_name,
        "line": call["site"]["line"],
        "column": call["site"]["column"],
        "endLine": call["endSite"]["line"],
        "endColumn": call["endSite"]["column"],
        "argument": argument,
        "shape": call["shape"],
        "targets": sorted(targets),
        "configuration": [
            {"path": path, "jsonPointer": pointer}
            for path, pointer in sorted(configuration)
        ],
        "entryPointGroup": group,
        "evidenceError": error,
    }
